src/logic/synchronization.ts

```diff
@@ -0,0 +1,106 @@
+import _ from "lodash";
+import "../utils/lodash-mixins";
+
+import Instance from "../models/instance";
+import { d2ModelFactory } from "../models/d2ModelFactory";
+import { D2 } from "../types/d2";
+import {
+    ExportBuilder,
+    NestedRules,
+    SynchronizationBuilder,
+    SynchronizationResult,
+} from "../types/synchronization";
+import { cleanObject } from "../utils/d2";
+import {
+    buildNestedRules,
+    getAllReferences,
+    getMetadata,
+    postMetadata,
+} from "../utils/synchronization";
+
+async function exportMetadata(d2: D2, builder: ExportBuilder): Promise<SynchronizationResult> {
+    const { type, ids, excludeRules, includeRules } = builder;
+    const model = d2ModelFactory(d2, type).getD2Model(d2);
+    const result: SynchronizationResult = {};
+
+    // Each level of recursion traverse the exclude/include rules with nested values
+    const nestedExcludeRules: NestedRules = buildNestedRules(excludeRules);
+    const nestedIncludeRules: NestedRules = buildNestedRules(includeRules);
+
+    // Get all the required metadata
+    const syncMetadata = await getMetadata(d2, ids);
+    const elements = syncMetadata[model.plural] || [];
+
+    // TODO: We are now doing a sequential synchronization
+    // Parallel requests could be done but we need to implement a Promise queue (cwait)
+    // Attempting to perform all the requests without a queue blocks the server and affects performance
+    for (const element of elements) {
+        const object = cleanObject(element, excludeRules);
+
+        // Store Organisation Unit in result object
+        result[model.plural] = result[model.plural] || [];
+        result[model.plural].push(object);
+
+        // Get all the referenced metadata
+        const references: SynchronizationResult = getAllReferences(d2, object, model.name);
+        const referenceTypes = _.intersection(_.keys(references), includeRules);
+        const promises = _.map(referenceTypes, type => {
```

Comment:
Maybe a normal `Array.map` here?

Reply:
As the block is an expression, it can be simplified: `.map(type => ({type: ....}))`

Reply:
Done and done